lib/explorer/series.ex

```diff
@@ -3428,7 +3428,19 @@ defmodule Explorer.Series do
   """
   @doc type: :element_wise
   @spec pow(left :: Series.t() | number(), right :: Series.t() | number()) :: Series.t()
-  def pow(left, right), do: basic_numeric_operation(:pow, left, right)
+  def pow(left, right) do
+    if K.or(match?(%Series{}, left), match?(%Series{}, right)) do
+      apply_series_list(:pow, [cast_for_pow(left), cast_for_pow(right)])
+    else
+      raise ArgumentError, "at least one input must be a series"
+    end
+  end
+
+  @non_finite [:nan, :infinity, :neg_infinity]
+  defp cast_for_pow(%Series{} = series), do: series
+  defp cast_for_pow(u) when K.and(is_integer(u), u >= 0), do: from_list([u], dtype: {:u, 64})
```

Comment:
I added the `cast_for_pow` function entirely for this clause. It allows you to get:

```elixir
iex> s1 = Explorer.Series.from_list([1, 2, 3])
iex> Explorer.Series.pow(s1, 2)
# #Explorer.Series<
#   Polars[3]
#   s64 [1, 4, 9]
# >
```

I.e. raising a signed int series to an unsigned int literal keeps the result dtype as signed int instead of casting to a float.

Per out conversation on slack about unsigned ints, I'm wondering if this actually desirable. It's at odds with:

```elixir
iex> s1 = Explorer.Series.from_list([1, 2, 3])
iex> s2 = Explorer.Series.from_list([2])
iex> Explorer.Series.pow(s1, s2)
# #Explorer.Series<
#   Polars[3]
#   f64 [1.0, 4.0, 9.0]
# >
```

If we didn't have that clause, though, you'd have to do this to keep the dtype as signed int:

```elixir
iex> s1 = Explorer.Series.from_list([1, 2, 3])
iex> s3 = Explorer.Series.from_list([2], dtype: :u64)
iex> Explorer.Series.pow(s1, s3)
# #Explorer.Series<
#   Polars[3]
#   s64 [1, 4, 9]
# >
```

which is less ergonomic than just `Explorer.Series.pow(s1, 2)`.

Thoughts?

Reply:
Could we, by any chance, have Nx behaviour? If integers are given, it raises if the right side is negative, and always returns integers. If you want a float response, you just cast any of the results to float. I guess this could be achieved by casting the right side to `{:u, precision}` always when a `{:s, precision}` is given?

Reply:
> Could we, by any chance, have Nx behaviour? If integers are given, it raises if the right side is negative, and always returns integers.

I think this would break with Polars since they allow `pow(uint, sint) == float`.

My reason for this cast was that our integers are "untyped" (no `s` vs. `u`). I tried to make the UX a little smoother by assuming any integer > 0 was unsigned. This breaks with our normal type inference, though.

> If you want a float response, you just cast any of the results to float. I guess this could be achieved by casting the right side to {:u, precision} always when a {:s, precision} is given?

Yeah a user can always achieve a more specific result by casting.

Reply:
@josevalim We could of course impose our own type restrictions if we want. We'd just need to pattern match on the dtypes.

Definitely doable :)